### PR TITLE
[Automation] Add CI linters/checks

### LINKED ISSUE #175: [Automation] Automate Go Formatting and similar tasks
## Objective

Create a foundational infrastructure so it is to automate certain tasks on pre-commits or PRs to avoid doing repetitive tasks.

## Origin Document

In https://github.com/pokt-network/pocket/pull/171, we did a code review to follow up on another PR after it is merged to master. See https://github.com/pokt-network/pocket/pull/166#issuecomment-1223262358 for details.

This approach raised the need for being able to easily automated tasks, ranging from code formatting, to tending to TODOs, running tests, etc...

## Goals

- [ ] Create a foundation so it is easy to add/automate new tasks that should be used and followed by others on the team
- [ ] Pick a technology that can cover most of our use cases (github workflows, git recommit hooks, etc...)

## Deliverable

- [ ] Document (1 page or less) on how we can add new automations
- [ ] At a minimum, automate the following:
    - [ ] Running `go fmt` so the imports are always ordered
    - [ ] Add an ability to run manual checks to make sure things like `require.True` are not used in the same line as `bytes.Equal` enforcing users to use `require.Equal`. Note that an override option will be necessary.
    - [ ] Remove unnecessary newlines

## Non-goals / Non-deliverables

- Making sure this solution is generalizable to anything
- Automating all the different things we need

## General issue deliverables

- [ ] Update the appropriate CHANGELOG
- [ ] Update any relevant READMEs (local and/or global)
- [ ] Update any relevant global documentation & references
- [ ] If applicable, update the source code tree explanation
- [ ] If applicable, add or update a state, sequence or flowchart diagram using [mermaid](https://mermaid-js.github.io/mermaid/)

---

**Creator**: @Olshansk 
**Co-Owners**: @okdas

### PR DESCRIPTION
**Pull request by okdas**
## Description

Adding linters and static checkers to CI rotation.

## Issue

CI part of https://github.com/pokt-network/pocket/issues/175

## Type of change

Please mark the options that are relevant.

- [ ] New feature, functionality or library
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Code health or cleanup
- [ ] Major breaking change
- [x] Documentation
- [ ] Other: <REPLACE_ME>

## List of changes

- Adding more checks for CI

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have tested my changes using the available tooling
- [x] If applicable, I have made corresponding changes to related local or global README
- [ ] If applicable, I have added new diagrams using [mermaid.js](https://mermaid-js.github.io)
- [ ] If applicable, I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: .github/workflows/main.yml
Comment (Olshansk) [inline .github/workflows/main.yml:66]: Can you cahnge `TODO` to `TODO(kdas)` here since you're the likely owner of it?

File: .github/workflows/main.yml
Comment (Olshansk) [inline .github/workflows/main.yml:74]: Why did we add `always()` here?

File: .github/workflows/linters.yml
Comment (Olshansk) [inline .github/workflows/linters.yml:3]: Could you add some documentation (in a README, in a comment, etc...) on what the next steps here are?

I have various open questions. E.g.:
1. When/who is responsible for updating the branches?
2. When/who is responsible for updating the paths?
3. When/who is responsible for adding linter checks?
4. Is there a repository of linter checks we reference or do we create our own?

No need to answer everything in line, but I'm just not sure what the next steps are.

**Review by Olshansk — CHANGES_REQUESTED** [review]
Please update the PR description.

**Comment by Olshansk** [comment]
@okdas We might be doing a techdebt "bug bash" next week. Do you think it'd be possible to have a baseline implementation of this that we can build on / add to?

**Comment by okdas** [comment]
> Do you think it'd be possible to have a baseline implementation of this that we can build on / add to?

Yes, if this is not going to be ready by then I will export linters output we can go through together.

**Comment by Olshansk** [comment]
@okdas Just wanted to follow up if there are any blockers on this PR?

**Comment by okdas** [comment]
> @okdas Just wanted to follow up if there are any blockers on this PR?

No blockers, will pick it up soon.

File: .github/workflows/main.yml
Comment (okdas) [reply .github/workflows/main.yml:74]: > Why did we add `always()` here?

`always()` forces step to run even if previous step(s) failed. 

https://docs.github.com/en/actions/learn-github-actions/expressions#always

Comment (Olshansk) [reply .github/workflows/main.yml:73]: Cool. Ty for the link and explanation!

File: docs/releases/README.md
Comment (Olshansk) [inline docs/releases/README.md:21]: Can you insert links to the images built for reference for each of the subsections below?

File: docs/releases/README.md
Comment (Olshansk) [inline docs/releases/README.md:23]:
```suggestion
Code built from the default branch (i.e. `main`) is tagged as `latest`.
```

File: docs/releases/README.md
Comment (Olshansk) [inline docs/releases/README.md:27]:
```suggestion
Once releases are managed, they will be tagged with the version number (e.g. `v0.0.1-alpha.pre.1`).
```

File: docs/releases/README.md
Comment (Olshansk) [inline docs/releases/README.md:32]:
```suggestion
Extended images with additional tooling are built to aid in troubleshoot and debugging. The extended image is formatted as `<tag>-dev`. For example, `latest-dev`, or `pr-123-dev`.
```

File: Makefile
Comment (Olshansk) [inline Makefile:322]:
```suggestion
## Run all linters that are triggered by the CI pipeline
```

File: docs/development/README.md
Comment (Olshansk) [inline docs/development/README.md:169]: Where possible/simple, can you add a make target that runs the commands to install them and update the docs here?

File: docs/development/README.md
Comment (Olshansk) [inline docs/development/README.md:161]: Please add a section explaining "how does one add new linters?"

**Review by Olshansk — COMMENTED** [review]
I think you can change this from a draft to a real PR.

**Comment by Olshansk** [comment]
@okdas Do you need another review on this or is it still a WIP?

File: docs/development/README.md
Comment (okdas) [reply docs/development/README.md:169]: @Olshansk I think it's better to keep the link to installation instructions. Snippet from `golangci-lint` docs:
> such `go install`/`go get` installation aren't guaranteed to work. We recommend using binary installation.

I'm not aware of other OS-agnostic way of installing the binary we could use in Makefile. :(

File: docs/development/README.md
Comment (okdas) [reply docs/development/README.md:166]: @Olshansk I've added more documentation, including the part on how to manage linters, and how to write in-house checks/rules.

File: docs/releases/README.md
Comment (okdas) [reply docs/releases/README.md:31]: @Olshansk instead of attaching an image, I've put a link to the page with list of tags which can be used as a reference. Note, this container repo is currently private, and we should make it public.

**Comment by okdas** [comment]
> @okdas Do you need another review on this or is it still a WIP?

this is ready for review now, thanks!

File: docs/development/README.md
Comment (Olshansk) [reply docs/development/README.md:186]: Screen Shot 2022-09-27 at 6 02 18 PM

Configured.

File: docs/development/README.md
Comment (Olshansk) [inline docs/development/README.md:168]:
```suggestion
We utilize `golangci-lint` to run the linters. It is a wrapper around a number of linters and is configured to run many at once. The linters are configured to run on every commit and pull request via CI, and all code issues are populated as GitHub annotations to let developers and reviewers easily locate an issue.
```

File: docs/development/README.md
Comment (Olshansk) [inline docs/development/README.md:172]: Make target that helps with installation?

File: docs/releases/README.md
Comment (Olshansk) [reply docs/releases/README.md:31]: Nice!

File: docs/development/README.md
Comment (Olshansk) [reply docs/development/README.md]: Cool. Thanks for looking into it!

File: docs/development/README.md
Comment (Olshansk) [reply docs/development/README.md:166]: LGTM!

File: docs/development/README.md
Comment (Olshansk) [inline docs/development/README.md:197]:
```suggestion
`golangci-lint` is a sophisticated tool including both default and custom linters. The configuration file, which can grow quite large, is located at [`.golangci.yml`](../../.golangci.yml).
```

File: docs/development/README.md
Comment (Olshansk) [inline docs/development/README.md:203]: cc @kutoft regarding this entire README.

Once it is merged to main, I will ping the team directly in discord, but there's no reason for your team not to leverage all this infra.

File: docs/releases/README.md
Comment (Olshansk) [inline docs/releases/README.md:38]:
```suggestion
```

File: docs/releases/README.md
Comment (Olshansk) [inline docs/releases/README.md:43]:
```suggestion
```

**Review by Olshansk — CHANGES_REQUESTED** [review]
Left a few some suggestions/comments throughout but looking forward to getting this in!

Very well designed and documented!

File: misc/linting-rules/tests.go
Comment (okdas) [reply misc/linting-rules/tests.go:1]: I think using `build` dir is more appropriate. I'll do that change.

File: .github/workflows/main.yml
Comment (okdas) [reply .github/workflows/main.yml:107]: @Olshansk prior to that change, we pushed the image to two repos: github container registry (ghcr.io) and docker hub. For some reason, docker hub doesn't always authenticate the user, hence failure to upload the image. We are going to be just fine with utilizing ghcr.io (and I'm thinking of possibly adding AWS ECR as a backup later).

**Comment by Olshansk** [comment]
@okdas Lmk when I should re-review here. 

Someone on the team will likely need to a single followup PR because it looks like we caught a bunch of stuff :)

Screen Shot 2022-09-29 at 5 50 16 PM

**Comment by okdas** [comment]
@Olshansk you can re-review now.

> Someone on the team will likely need to a single followup PR because it looks like we caught a bunch of stuff :)

Haha, yes!

Here's an example of `golangci-lint run --fix` changes: https://github.com/pokt-network/pocket/compare/dk-add-linters...dk-add-linters-autofix?expand=1 - some of the issues can be fixed automatically.

Note: we have a very limited number of linters turned on! If you want to go crazy you could run `golangci-lint run --enable-all --fix` - just beware this changes lots of files, some linters can be redundant, and in order to run this command you'd need to comment out `enabled: []` section in `.golangci.yml` file.

File: Makefile
Comment (Olshansk) [inline Makefile:329]: Can you move this next to the other `go_*` targets?

Right now its just in between all the unit tests.

Screen Shot 2022-10-03 at 3 10 44 PM

**Review by Olshansk — APPROVED** [review]
@okdas 

1. Next time, please make sure to "re-request review" when something is ready to be re-reviewed:

Screen Shot 2022-10-03 at 3 08 48 PM

2. I left one open comment. please update it before merging

Screen Shot 2022-10-03 at 3 10 44 PM

**Pull request merged**
